### PR TITLE
Resolving #2153  Added ability to set frame-lock.

### PR DESCRIPTION
**Pull request by TheoTheDev**
Added ability to increase / decrease max FPS.
By default disabled.
By pressing shift & -  -> decrease the limit.
By pressing shift & + -> increase the limit.
Also updated stats FPS with render FPS & loopFPS:
image
+ in console you can see what exact limit was set [after changing the limit].

### CODE REVIEW

File: src/components/app/App.jsx [hunk 1]
```diff
@@ -141,6 +141,36 @@ export const App = () => {
 
     }, [] );
 
+    useEffect( () => {
+
+        const handleKeyUp = ( event ) => {
+
+            if ( event.which === 187 && event.shiftKey ) {
```
Comment (0reo) [inline src/components/app/App.jsx:148]: please add a comment with which key this is for readability

File: src/components/app/App.jsx [hunk 2]
```diff
@@ -141,6 +141,36 @@ export const App = () => {
...
+
+                app.renderLimitMs = app.renderLimitMs || 16 * 2; // if not set then set to 30fps
+                app.renderLimitMs = Math.floor( 1000 / ( 1000 / app.renderLimitMs + 5 ) ); // +5fps, can be not 100% accurate cause of update/render frame shifted [loop time too long]
+                console.log( `Render limit increased to ${ app.renderLimitMs }ms (${ Math.floor( 1000 / app.renderLimitMs ) }fps)` );
+
+            } else if ( event.which === 189 && event.shiftKey ) {
```
Comment (0reo) [inline src/components/app/App.jsx:154]: please add a comment with which key this is for readability

**Review by 0reo — CHANGES_REQUESTED** [review]
Few readability changes would be good

**Comment by TheoTheDev** [comment]
Added comments as requested + added check is typing inside of input [in this case don't apply].

**Review by 0reo — COMMENTED** [review]
should be good for testing @alisaad673

File: src/components/app/App.jsx [hunk 3]
```diff
@@ -141,6 +141,43 @@ export const App = () => {
 
     }, [] );
 
+    useEffect( () => {
```
Comment (avaer) [inline src/components/app/App.jsx:144]: This is not a good place to hook in the functionality. It sounds more like a job for `ioManager`.

Comment (TheoTheDev) [reply src/components/app/App.jsx]: Updated and moved key handling to ioManager.

**Comment by alisaad673** [comment]
Decreasing and increasing the limit seems to be working fine but render fps don't go be beyond 60 fps even if the limit is set at 100fps. Is it dependent on refresh rate of screen?

https://user-images.githubusercontent.com/35926530/163037376-e26e4acb-55bc-4556-b91d-17f602216418.mp4

**Comment by TheoTheDev** [comment]
> Decreasing and increasing the limit seems to be working fine but render fps don't go be beyond 60 fps even if the limit is set at 100fps. Is it dependent on refresh rate of screen?
> 
>  bandicam.2022-04-13.00-15-51-603.mp4

Usually max refresh rate is 60fps. But if your device has low performance and you get max 40fps for example you can not go beyond this physical limit of computation power of your device.

**Comment by avaer** [comment]
>Usually max refresh rate is 60fps.

This should be determined by the browser and how fast it fires `requestAnimationFrame`. We should not hardcode any values here.

**Comment by TheoTheDev** [comment]
> > Usually max refresh rate is 60fps.
> 
> This should be determined by the browser and how fast it fires `requestAnimationFrame`. That is not necessarily 60FPS.
> 
> (note that we should not hardcode any values here)

so anything to be changed ?
you want to replace 2 * 16.6 with a constant and that's it ?
anything else ?

File: webaverse.js [hunk 4]
```diff
@@ -71,6 +71,10 @@ export default class Webaverse extends EventTarget {
   constructor() {
     super();
 
+    this.lastRenderTime = 0;
+    this.renderFrame = 0;
```
Comment (avaer) [inline webaverse.js:75]: These are port variable names. Is it the number of frames?

File: io-manager.js [hunk 5]
```diff
@@ -570,6 +570,30 @@ ioManager.keydown = e => {
       debug.toggle();
       break;
     }
+    case 187: { // + & shift [plus - not on num-pad]
```
Comment (avaer) [inline io-manager.js:573]: Rather than this keyboard hack, this should be in the settings.

**Comment by avaer** [comment]
This isn't likely to ship.

**Pull request closed without merging**